metrics.py

```diff
@@ -67,41 +67,40 @@ def document_count_by(
         column: str,
         var_name: str) -> DataFrame:
     """Count documents by grouping column."""
-    # out = df.withColumn('datasourceId', lit('all'))
     out = df.groupBy(column).count().alias('count')
     out = out.withColumn('variable', f.lit(var_name))
     out = out.withColumn('field', f.lit(None).cast(t.StringType()))
     return out
 
 
-def evidence_not_null_fields_count(
-        df: DataFrame,
-        var_name: str) -> DataFrame:
-    """Count number of evidences with not null values in variable."""
-    # flatten dataframe schema
+def not_null_fields_count(df: DataFrame, var_name: str, group_by_datasource: bool) -> DataFrame:
+    """Count number of not null values for each field in the dataframe. If `group_by_column` is provided, the
+    calculation is performed separately for every distinct value in that column."""
+    # Flatten the dataframe schema.
     flat_df = df.select([f.col(c).alias(c) for c in flatten(df.schema)])
-
-    # counting not-null evidence per field
-    exprs = [f.sum(f.when(f.col(field.name).getItem(0).isNotNull(), f.lit(1))
-                    .otherwise(f.lit(0))).alias(field.name)
-             if isinstance(field.dataType, t.ArrayType)
-             else
-             f.sum(f.when(f.col(field.name).isNotNull(), f.lit(1))
-                    .otherwise(f.lit(0))).alias(field.name)
-             for field in list(filter(lambda x: x.name != 'datasourceId',
-                                      flat_df.schema))]
-    out = df.groupBy(f.col('datasourceId')).agg(*exprs)
-    # Clean column names
-    out_cleaned = out.toDF(*(c.replace('.', '_') for c in out.columns))
-    # wide to long format
-    cols = out_cleaned.drop('datasourceId').columns
-
-    melted = melt(out_cleaned,
-                  id_vars=['datasourceId'],
-                  var_name='field',
-                  value_vars=cols,
-                  value_name='count')
+    # Get the list of fields to count.
+    field_list = flat_df.schema
```

Comment:
```suggestion
    field_list = flat_df.columns if not group_by_datasource else flat_df.drop('datasourceId').columns
```

This would remove lines 82-84. Also, do you use `.schema` because you need a list of the datatypes? If you simply need a list of the columns, .`columns` works just like in Pandas :)

Reply:
Oh, you're absolutely right! The use of `.schema` was there before, and I just refactored it to be the same without thinking. This is indeed much more elegant; addressed in https://github.com/opentargets/ot-release-metrics/pull/4/commits/cadd7a0ad8ce1bf4c090a78fc364e399de254bc7.

Reply:
Whoops, actually no, this breaks things. In calculating `exprs`, we check `if isinstance(field.dataType, t.ArrayType)`, so we can't have the field be just a string. Sorry, I also didn't notice this. Reverted in https://github.com/opentargets/ot-release-metrics/pull/4/commits/73a4bc8a583bcb0f591cf32f6fea631017161136